pycqed/measurement/waveform_control/segment.py

```diff
@@ -810,15 +829,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
+                              f'starts at {el_new_start*1e6:.4f}us) overlap ' \
+                              f'on {awg}'
+                        if allow_overlap:
+                            log.warning(msg)
+                        else:
+                            raise ValueError(msg)
+
+    def resolve_overlap(self):
+        """
+        Routine to resolve overlapping elements. Will be exectued if
+        corresponding pulsar parameter is true. This code
+        first goes through the list of overlapping elements that are
+        pairwise overlapping and clusters them into lists
+        of overlapping elements, where two different lists of overlapping
+        elements have no overlapping elements with
+        one another. At the end the code combines all elements of each
+        list into a new element.
+        """
+
+        self.gen_elements_on_awg()
+        self._test_overlap(track_and_ignore=True)
+        overlapping_elements = self.overlapping_elements
+
+        if len(overlapping_elements)==0:
+            return
+
+        # add first two overlapping elements to list
+        joint_overlapping_elements = [overlapping_elements[0]]
+
+        new_cluster = True
+        for i in range(len(overlapping_elements) - 1):
+            # check for all sets added to joint_overlapping_elements
+            # whether the next set of elements from
+            # overlapping_elements shares an element name. If so,
+            # then add this element to the entry in
+            # joint_overlapping_elements
+            for j in range(len(joint_overlapping_elements)):
+                if len(joint_overlapping_elements[j] & \
+                       overlapping_elements[i + 1]) != 0:
+                    joint_overlapping_elements[j] = \
+                        joint_overlapping_elements[j] | \
+                        overlapping_elements[i + 1]
+                    new_cluster = False
+
+            # if the new element from overlapping_elements overlaps
+            # with none of the previously added elements in
+            # joint_overlapping_elements (i.e. if new_cluster=True)
+            # add it as a new cluster.
+            if new_cluster:
+                joint_overlapping_elements.append(overlapping_elements[i + 1])
+            new_cluster = True
+
+        for i in range(len(joint_overlapping_elements)):
+            self._combine_elements(joint_overlapping_elements[i],
+                                   'overlapping_el_{}_{}'.format(i, self.name))
+
+
+    def _combine_elements(self, elements, combined_el_name):
+        """
+        Routine to properly combine elements in the segment.
+        :param elements: list or set of elements in the segment to be combined
+        :param combined_el_name: name of the combined element
+        :return:
+        """
+
+        new_pulse_list = []
+
+        for el in elements:
+            new_pulse_list+= self.elements.pop(el)
```

Comment:
Shouldn't you update `p.pulse_obj.element_name` to the new name for p in new_pulse_list?

Reply:
I cannot see it unfortunately. Maybe we can briefly talk.

Reply:
Each entry in your new_pulse_list is an UnresolvedPulse object, which has a pulse_obj property, see the init of UnresolvedPulse. This pulse_obj is a Pulse object, and thus has a property element_name, see the init of Pulse in pycqed.measurement.waveform_control.pulse. This property element_name will be outdated after your code.

Reply:
Ah yes, now I see it and makes sense! (self.elements is however a list of pulse objects already after resolving, so updating the p.element_name directly is sufficient).